File: resources/ToArray.cpp
#include <fstream>

int main(int argc, char *argv[])
{
	if (argc != 5)
	{
		return 1;
	}
	auto *outputCppPath = argv[1];
	auto *outputHPath = argv[2];
	auto *inputAnyPath = argv[3];
	auto *symbolName = argv[4];
	std::ifstream inputAny(inputAnyPath, std::ios::binary);
	std::ofstream outputCpp(outputCppPath);
	if (!outputCpp)
	{
		return 2;
	}
	outputCpp << "#include \"" << outputHPath << "\"\nconst unsigned char " << symbolName << "[] = { ";
	auto dataLen = 0U;
	while (true)
	{
		char ch;
		inputAny.read(&ch, 1);
		if (inputAny.eof())
		{
			break;
		}
		if (!inputAny)
		{
			return 3;
		}
		outputCpp << (unsigned int)(unsigned char)(ch) << ", ";
		dataLen += 1;
	}
	outputCpp << " }; const unsigned int " << symbolName << "_size = " << dataLen << ";\n";
	if (!outputCpp)
	{
		return 4;
	}
	std::ofstream outputH(outputHPath);
	if (!outputH)
	{
		return 5;
	}
	outputH << "#pragma once\nextern const unsigned char " << symbolName << "[]; extern const unsigned int " << symbolName << "_size;\n";
	if (!outputH)
	{
		return 6;
	}
	return 0;
}

File: resources/to-array.py
import os
import sys

(
	script,
	output_cpp_path,
	output_h_path,
	output_dep_path,
	input_path,
	symbol_name,
) = sys.argv

script_path = os.path.realpath(__file__)

with open(input_path, 'rb') as input_f:
	data = input_f.read()
data_size = len(data)
bytes_str = ', '.join([ str(ch) for ch in data ])

with open(output_cpp_path, 'w') as output_cpp_f:
	output_cpp_f.write(f'''
#include "{output_h_path}"

const struct {symbol_name}Resource {symbol_name} = {{{{{{ {bytes_str} }}}}}};
''')

with open(output_h_path, 'w') as output_h_f:
	output_h_f.write(f'''
#pragma once
#include <array>
#include <span>

extern const struct {symbol_name}Resource
{{
	std::array<unsigned char, {data_size}> data;

	std::span<const char> AsCharSpan() const
	{{
		return std::span(reinterpret_cast<const char *>(data.data()), data.size());
	}}

	std::span<const unsigned char> AsUcharSpan() const
	{{
		return std::span(data.data(), data.size());
	}}
}} {symbol_name};
''')

def dep_escape(s):
	t = ''
	for c in s:
		if c in [ ' ', '\\', ':', '$' ]:
			t += '\\'
		t += c
	return t

with open(output_dep_path, 'w') as output_dep_f:
	output_dep_f.write(f'''
{dep_escape(output_cpp_path)} {dep_escape(output_h_path)}: {dep_escape(input_path)} {dep_escape(script_path)}
''')
